### ISSUE TITLE
Self-referential Data Modeling Leads to Unexpected Behavior

### ISSUE THREAD

**Issue by mikeslade** [prisma/prisma2#291]
**Problem:**
The order of field elements in a model impact whether the self-referential relationship behaves appropriately.

**Environment (Docker containers):**
- Prisma2 pre-release 3.2 running on Ubuntu (latest)
- Postgres database

**Example:**
I have a model Group, that can have many child Groups. 

_First Attempt_
Simply applying the reference to a single field, like the example from the docs (https://github.com/prisma/prisma2/blob/master/docs/relations.md#self-relations) leads to the following error when migrating via `prisma2 lift up`: "Ambiguous self relation detected."

```
model Group {
  id    String  @default(cuid()) @id @unique
  name String
  parent Group?
}
```

_Second Attempt_
So, if I then explicitly create a parent field for Group and a childGroups field for Group[] on the model with a reference linking them, this clears up the lift up error, but then leads to unexpected outcomes when running Photon.

```
model Group {
  id    String  @default(cuid()) @id @unique
  name String
  parent Group? @relation(name: "ChildGroups")
  childGroups Group[] @relation(name: "ChildGroups")
}
```

If I try to create a parent group with its children in one pass, like this:

```
await photon.groups.create({
      data: {
        name: "PARENT_GROUP",
        childGroups: {
          create: [
            {
              name: "CHILD_GROUP1",
            },
            {
              name: "CHILD_GROUP2",
            },
            {
              name: "CHILD_GROUP3",
            }
          ]
        }
      }
})
```

Or if I try to create each group one by one, like this:

```
const parent = await photon.groups.create({
    data: {
      name: "PARENT_GROUP",
    }
  });

  if (parent) {
    const group1 = await photon.groups.create({
      data: {
        name: "CHILD_GROUP1",
        parent: {
          connect: {
            id: parent.id
          }
        },
      }
    });

    const group2 = await photon.groups.create({
      data: {
        name: "CHILD_GROUP2",
        parent: {
          connect: {
            id: parent.id
          }
        },
      }
    });

    const group3 = await photon.groups.create({
      data: {
        name: "CHILD_GROUP3",
        parent: {
          connect: {
            id: parent.id
          }
        },
      }
    });
  }
```

They both result with the parent group being the only one with a reference in the parent field and it is to the last child group created.

id | name | parent
cjyox9y2c0002alrzxi4wvgf1 | CHILD_GROUP1 | null
cjyox9y2q0003alrz723c94uc | CHILD_GROUP2 | null
cjyox9y320004alrzmyl6d1yv | CHILD_GROUP3 | null
cjyox9y200001alrzqotpza6d | PARENT_GROUP | cjyox9y320004alrzmyl6d1yv

I am looking for the opposite behavior, where each child group has a reference to its parent if there is one. 

_Final Attempt - Success_
Simply ordering the childGroups field before the parent field resulted in the desired behavior:

```
model Group {
  id    String  @default(cuid()) @id @unique
  name String
  childGroups Group[] @relation(name: "ChildGroups")
  parent Group? @relation(name: "ChildGroups")
}
```

id | name | parent
cjyox9y200001alrzqotpza6d | PARENT_GROUP | null
cjyox9y2c0002alrzxi4wvgf1 | CHILD_GROUP1 | cjyox9y200001alrzqotpza6d
cjyox9y2q0003alrz723c94uc | CHILD_GROUP2 | cjyox9y200001alrzqotpza6d
cjyox9y320004alrzmyl6d1yv | CHILD_GROUP3 | cjyox9y200001alrzqotpza6d

**Expected Behavior:**
Order of model fields would not matter and still result in the same final output.

**Comment by pantharshit00** [comment]
I am even unable to deploy this datamodel in SQLite. Possibly related to #200

**Comment by mavilein** [comment]
@pantharshit00 : Can you test that again on latest alpha please? I worked in that area today and might have fixed this.

**Comment by pantharshit00** [comment]
@mavilein Still unable to deploy with `prisma2@2.0.0-alpha.76, binary version: 308a79c48cb22db008241f79742f20803cee710b`

# Failed listMigrations at 2019-07-31T17:17:24.198Z
## RPC Input One Line
```json
{"id":1,"jsonrpc":"2.0","method":"listMigrations","params":{"projectInfo":"","sourceConfig":"datasource db {\n  provider = \"sqlite\"\n  url      = \"file:dev.db\"\n  default  = true\n}\n\ngenerator photon {\n  provider = \"photonjs\"\n}\n\nmodel Group {\n  id   String @default(cuid()) @id @unique\n  name String\n   parent Group? @relation(name: \"ChildGroups\")\n  childGroups Group[] @relation(name: \"ChildGroups\")\n}\n}"}}
```

## RPC Input Readable
```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "listMigrations",
  "params": {
    "projectInfo": "",
    "sourceConfig": "datasource db {\n  provider = \"sqlite\"\n  url      = \"file:dev.db\"\n  default  = true\n}\n\ngenerator photon {\n  provider = \"photonjs\"\n}\n\nmodel Group {\n  id   String @default(cuid()) @id @unique\n  name String\n   parent Group? @relation(name: \"ChildGroups\")\n  childGroups Group[] @relation(name: \"ChildGroups\")\n}\n}"
  }
}
```


## RPC Response
```
null
```

## Stack Trace
```bash
thread 'main' panicked at 'loading the connector failed.: DataModelErrors { code: 1001, errors: ["Unexpected token. Expected one of: end of input, type declaration, model declaration, enum declaration, source definition, generator definition."] }', src/libcore/result.rs:997:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::_print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::continue_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: migration_core::migration_engine::MigrationEngine::init
  10: <F as jsonrpc_core::calls::RpcMethodSimple>::call
  11: <F as jsonrpc_core::calls::RpcMethod<T>>::call
  12: <futures::future::lazy::Lazy<F,R> as futures::future::Future>::poll
  13: <futures::future::then::Then<A,B,F> as futures::future::Future>::poll
  14: <futures::future::map::Map<A,F> as futures::future::Future>::poll
  15: <futures::future::either::Either<A,B> as futures::future::Future>::poll
  16: futures::task_impl::std::set
  17: std::thread::local::LocalKey<T>::with
  18: futures::future::Future::wait
  19: jsonrpc_core::io::IoHandler<M>::handle_request_sync
  20: migration_core::rpc_api::RpcApi::handle
  21: migration_engine::main
  22: std::rt::lang_start::{{closure}}
  23: std::panicking::try::do_call
  24: __rust_maybe_catch_panic
  25: std::rt::lang_start_internal
  26: main

```

**Comment by mavilein** [comment]
@pantharshit00 : Please try again now. Lift should not fail anymore.

**Comment by pantharshit00** [comment]
Ok, I am finally able to deploy the datamodel to SQLite.

Now that is possible, I am able to successfully reproduce this issue posted above.

**Comment by mavilein** [comment]
This is fixed in binary version `b8957df02bca6708b7aefa56832eda409dbb1114`. It is already available on `alpha`.
